components/operator/internal/controllers/stack/stack_finalizer.go

```diff
@@ -0,0 +1,299 @@
+}
+
+func (f *StackFinalizer) HandleFinalizer(ctx context.Context, log logr.Logger, stack *stackv1beta3.Stack, conf *stackv1beta3.Configuration, req ctrl.Request) (*ctrl.Result, error) {
+
+	// examine DeletionTimestamp to determine if object is under deletion
+	// The object is being deleted
+	if !stack.ObjectMeta.DeletionTimestamp.IsZero() && controllerutil.ContainsFinalizer(stack, f.name) {
+		// Make sur that the object is disable
+		stack.Spec.Disabled = true
+		if err := f.client.Update(ctx, stack); err != nil {
+			return &ctrl.Result{}, err
+		}
+
+		// We also need to make sure that all deployements are Terminated,
+		// So that no one is already accessing our database.
+		found := &appsv1.DeploymentList{}
+		opts := &client.ListOptions{
+			Namespace: req.Name,
+		}
+
+		if err := f.client.List(ctx, found, opts); err != nil {
+			return &ctrl.Result{}, err
+		}
+
+		if len(found.Items) > 0 {
+			return &ctrl.Result{
+				Requeue:      true,
+				RequeueAfter: time.Second,
+			}, nil
+		}
+
+		// our StackFinalizer is present, so lets handle any external dependency
+		res, err := f.deleteStack(ctx, stack, conf, log)
+		if err != nil {
+			// if fail to delete the external dependency here, return with error
+			// so that it can be retried
+			return res, err
+		}
+
+		// remove our StackFinalizer from the list and update it.
+		controllerutil.RemoveFinalizer(stack, f.name)
+		if err := f.client.Update(ctx, stack); err != nil {
+			return &ctrl.Result{}, err
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return &ctrl.Result{}, nil
+
+	}
+
+	// The object is not being deleted, so if it does not have our StackFinalizer,
+	// then lets add the StackFinalizer and update the object. This is equivalent
+	// registering our StackFinalizer.
+	if !controllerutil.ContainsFinalizer(stack, f.name) {
+		controllerutil.AddFinalizer(stack, f.name)
+		if err := f.client.Update(ctx, stack); err != nil {
+			return &ctrl.Result{}, err
+		}
+	}
+
+	return nil, nil
+}
+
+var (
+	natsClientId = "membership"
+)
+var (
+	ErrCast = errors.New("cannot cast interface to string")
+)
+
+func (f *StackFinalizer) DeleteByBrokers(c *v1beta3.Configuration, stackName string, subjectService []string, logger logr.Logger) error {
+	values := reflect.ValueOf(c.Spec.Broker)
+	for i := 0; i < values.NumField(); i++ {
+		if rv := reflect.ValueOf(values.Field(i)); !rv.IsValid() || rv.IsNil() {
+			continue
+		}
+		switch t := values.Field(i).Interface().(type) {
+		case v1beta3.NatsConfig:
+			if err := f.deleleNatsSubjects(&t, stackName, subjectService, logger); err != nil {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+func (f *StackFinalizer) deleleNatsSubjects(config *v1beta3.NatsConfig, stackName string, subjectService []string, logger logr.Logger) error {
+	client, err := nats.NewClient(config, natsClientId)
+	if err != nil {
+		logger.Error(err, "NATS: client")
+		return err
+	}
+	defer client.Close()
+
+	jsCtx, err := client.JetStream()
```

Comment:
Maybe this client has to be closed?

Reply:
This is the context of Nats Jetstream, seems not being related to connection as i don't see any `.Close` or any method in the sdk. The client is already close one line above